### ISSUE TITLE
Passing rules in an array always returns 401

### ISSUE THREAD

**Issue by X-Tender** [tuupola/slim-jwt-auth#84]
I would like to redirect the user to a page when the error callbeck gets called.
Currently it's not possible because a 401 status gets attacked to the response (which is ok).
So currently it's not possible, right?

Maybe a feature for the next version to set an option of the 401 status gets added automaticly or not.

Or is there a workaround for this?

```php
"error" => function ($request, $response, $arguments) use ($container) {
    return $response->withRedirect( "my/target/path" );
},
```

**Comment by tuupola** [comment]
I think you can already do this with [3.x](https://github.com/tuupola/slim-jwt-auth/tree/3.x). There is no final release yet since PSR-15 has not been officially approved. However you can install it with:

```
$ composer require tuupola/slim-jwt-auth:3.0.0-rc.3
```

Be sure to check [CHANGELOG](https://github.com/tuupola/slim-jwt-auth/blob/3.x/CHANGELOG.md) and [README](https://github.com/tuupola/slim-jwt-auth/blob/3.x/README.md) since some things have changed.

**Comment by X-Tender** [comment]
OK, i give it a try and report later.

**Comment by X-Tender** [comment]
I've tested it out and get the same problem. Looks like because your ResponseFactory creates a response with 401 the slim framework didn't redirect.

Also I had problems setting the rules, when I made it this way:
```php
"path" => "/backend",
"ignore" => "/backend/login",
```
it worked, but this way it doesn't:
```php
"rules" => [
    new RequestPathRule([
        "path" => ["/backend"],
        "ignore" => ["backend/login"],
    ]),
    new RequestMethodRule([
        "ignore" => ["OPTIONS"],
    ])
]
```
I get the 401 error also on `"/"` route

**Comment by tuupola** [comment]
Have you tried setting the wanted status code in the error handler?

**Comment by tuupola** [comment]
I just made a quick test.

```php
<?php

require __DIR__ . "/vendor/autoload.php";

$app = new \Slim\App;

$app->add(new Tuupola\Middleware\JwtAuthentication([
    "secret" => "supersecretkeyyoushouldnotcommittogithub",
    "path" => "/backend",
    "ignore" => "/backend/login",
    "error" => function ($request, $response, $arguments) {
        return $response->withRedirect("/foo");
    }
]));

$app->get("/", function ($request, $response, $arguments) {
    return $response->write("home");
});

$app->get("/backend", function ($request, $response, $arguments) {
    return $response->write("backend");
});

$app->get("/backend/login", function ($request, $response, $arguments) {
    return $response->write("backend/login");
});

$app->run();
```

Doing a request which fails authentication returns a 302 response.

```
$ curl --include --insecure https://192.168.50.10/backend
HTTP/1.1 302 Found
Date: Tue, 16 Jan 2018 09:24:16 GMT
Server: Apache/2.2.15 (CentOS)
X-Powered-By: PHP/7.1.11
Location: /foo
```

Changing to 301 can be done in the error handler.

```php
$app->add(new Tuupola\Middleware\JwtAuthentication([
    "secret" => "supersecretkeyyoushouldnotcommittogithub",
    "path" => "/backend",
    "ignore" => "/backend/login",
    "error" => function ($request, $response, $arguments) {
        return $response->withRedirect("/foo", 301);
    }
]));
```

```
$ curl --include --insecure https://192.168.50.10/backend
HTTP/1.1 301 Moved Permanently
Date: Tue, 16 Jan 2018 09:25:31 GMT
Server: Apache/2.2.15 (CentOS)
X-Powered-By: PHP/7.1.11
Location: /foo
```

Will make a test about the rules bit later.

**Comment by tuupola** [comment]
Just noticed you are actually hit with a Slim 3.9.2 change. You must provide status code with `$response->withRedirect()` call or it will stay as 401.

**Comment by X-Tender** [comment]
Ok the 301 in the Redirect worked.
Can you maybe tell something about the "rules" problem I have?

**Comment by X-Tender** [comment]
OK '#ve figured it out. This way it works:
```php
$jwtAuth = (new JwtAuthentication(
    [
        "secret" => "secret",
        "secure" => "secure",
        "cookie" => "cookie",
        "error" => function ($request, $response, $arguments) use ($container) {
            return $response->withRedirect( $container->get("router")->pathFor("index"),301 );
        },
    ]
))->withRules([
    new RequestPathRule([
        "path" => "/lookbooks",
        "ignore" => ["/backend/login"],
    ]),
    new RequestMethodRule([
        "ignore" => ["OPTIONS"],
    ])
]);
$app->add($jwtAuth);
```
is the "rules" option deprecated or doesn't it work as expected? :)

**Comment by tuupola** [comment]
It should still work. It is a bug.

**Comment by tuupola** [comment]
Fixed in `3.0.0-rc.5`.

**Comment by contactjavas** [comment]
I face this (or similar 401) issue in the latest Slim (3.11.0) @tuupola
